src/server.cpp

```diff
@@ -20,19 +23,58 @@ void Server::start_receive()
         std::bind(&Server::handle_receive, this, pls::_1, pls::_2));
 }
 
-void Server::handle_receive(const error_code& error, std::size_t /*bytes_transferred*/)
+void Server::handle_receive(const error_code& error, std::size_t bytes)
 {
-    if (!error || error == boost::asio::error::message_size)
+    if (error)
     {
-        const std::string message("Hello, World!");
+        std::cout << "Error async_receivw_from: " << error.message() << "/n";
+        return;
+    }
+
+    if (bytes < 20)
+    {
+        std::cout << "Error: request length less than 20 bytes/n";
+        return;
+    }
 
-        socket_.async_send_to(boost::asio::buffer(message), remote_endpoint_,
-          std::bind(&Server::handle_send, this, message,pls::_1, pls::_2));
+    size_t length = recv_buffer_[2] * 256 + recv_buffer_[3];
 
-        start_receive();
+    if (bytes < length)
+    {
+        std::cout << "Error: request length " << bytes << "is less than specified in the request - " << length << "/n";
+        return;
     }
+
+    if (recv_buffer_[0] == ACCESS_REQUEST)
+        send_buffer_[0] = ACCESS_ACCEPT;
+    else
+        send_buffer_[0] = ACCESS_REJECT;
+
+    send_buffer_[1] = recv_buffer_[1];
+    send_buffer_[2] = 0;
+    send_buffer_[3] = 20;
+
+    for (size_t i = 0; i < 16; ++i)
+        send_buffer_[i + 4] = recv_buffer_[i + 4];
+
+    std::string secr("secret");
+
+    for (size_t i = 0; i < 6; ++i)
+        send_buffer_[i + 20] = secr[i];
+
+    std::array<uint8_t, 16> md;
+
+    MD5(send_buffer_.data(), 20 + secr.length(), md.data());
+
+    for (size_t i = 0; i < 16; ++i)
+        send_buffer_[i + 4] = md[i];
+
+    socket_.async_send_to(boost::asio::buffer(send_buffer_), remote_endpoint_,
```

Comment:
You are sending the whole `send_buffer_`, but you have only 20 bytes of data in it. All other bytes are garbage!

Don't send garbage, send just the data you need.